vignettes/advanced.Rmd

```diff
@@ -7,21 +7,67 @@ vignette: >
-demonstrate also carries over to `epix_slide()`.
+In this vignette, we discuss how to use the sliding functionality in the
+`epiprocess` package with less common grouping schemes or with computations that
+have advanced output structures.
+The output of a slide computation should either be an atomic value/vector, or a
+data frame. This data frame can have multiple columns, multiple rows, or both.
+
+During basic usage (e.g., when all optional arguments are set to their defaults):
+
+* `epi_slide(edf, <computation>, .....)`:
+
+  * keeps **all** columns of `edf`, adds computed column(s)
+  * outputs one row **for every row in `edf`** (recycling outputs from
+    computations appropriately if there are multiple time series bundled
+    together inside any group(s))
+  * maintains the grouping or ungroupedness of `edf`
+  * is roughly analogous to (the non-sliding) **`dplyr::mutate` followed by
+    `dplyr::arrange(time_value, .by_group = TRUE)`**
+  * outputs an **`epi_df`** if the required columns are present, otherwise a
+    tibble
+
+* `epix_slide(ea, <computation>, .....)`:
+
+  * keeps **grouping and `time_value`** columns of `ea`, adds computed
+    column(s)
+  * outputs one row **for element/row output from the computations**
+  * maintains the grouping or ungroupedness of `ea`, unless it was explicitly
+    grouped by zero variables; this isn't supported by `grouped_df` and it will
+    automatically turn into an ungrouped tibble
+  * is roughly analogous to (the non-sliding) **`dplyr::group_modify`**
+  * outputs a **tibble**
+
+These differences in basic behavior make some common slide operations require less boilerplate:
+* predictors and targets calculated with `epi_slide` are automatically lined up
+  with each other and with the signals from which they were calculated; and
+* computations for an `epix_slide` can output data frames with any number of
+  rows, containing models, forecasts, evaluations, etc., and will not be
+  recycled.
+
+When using more advanced features, more complex rules apply:
+
+* Generalization: `epi_slide(edf, ....., ref_time_values=my_ref_time_values)`
+  will output one row for every row in `edf` with `time_value` appearing inside
+  `ref_time_values`, and is analogous to a `dplyr::mutate`&`dplyr::arrange`
+  followed by `dplyr::filter` to those `ref_time_values`. We call this property
+  **size stability**, and describe how it is achieved in the following sections.
+  The default behavior described above is a special case of this general rule
+  based on a default value of `ref_time_values`.
+* Exception/feature: `epi_slide(edf, ....., ref_time_values=my_ref_time_values,
+  all_rows=TRUE)` will not just output rows for the given `ref_time_values`, but
+  instead will output one row per row in `edf`.
+* Exception/feature: `epix_slide(ea, ....., as_list_col=TRUE)` will output one
+  row per computation that outputs a data frame, even when these data frames
+  have arbitrary numbers of rows.
```

Comment:
This is only for `as_list_col = TRUE`.
```r
epix_slide(xx, f = function(x, g, t) data.frame(a = c(1, 2), b = "yes"),
             before = 2,
             new_col_name = "sum_binary",
             as_list_col = TRUE)
#> # A tibble: 4 × 2
#>   time_value sum_binary  
#> *      <dbl> <list>      
#> 1          4 <df [2 × 2]>
#> 2          5 <df [2 × 2]>
#> 3          6 <df [2 × 2]>
#> 4          7 <df [2 × 2]>

epix_slide(xx, f = function(x, g, t) c(1, 2),
             before = 2,
             new_col_name = "sum_binary",
             as_list_col = TRUE)
#> # A tibble: 8 × 2
#>   time_value sum_binary
#> *      <dbl>      <dbl>
#> 1          4          1
#> 2          4          2
#> 3          5          1
#> 4          5          2
#> 5          6          1
#> 6          6          2
#> 7          7          1
#> 8          7          2
```
The latter is a bug, but rather than fix I'd rather just eliminate the `as_list_col` parameter and have the user wrap their computation output in a list if they want this behavior, just like you would do in dplyr.  Thoughts?

Reply:
Ended up trying to fix the bug, and updated the wording here; hopefully it's more clear.

Reply:
Oooh, missed the `as_list_col=TRUE`. Makes sense now.

Reply:
I may have broken things with this implementation "fix" trying to make the wording easier.

Reply:
(But if we get this "fix" working it may auto-fix the problems with `unlist` that you brought up.)

Reply:
(One) problem with fix is not changing to `vctrs` stuff here:
```{r}
    if (all_rows) {
      orig_values = slide_values
      slide_values = rep(NA, nrow(.data_group))
      slide_values[o] = orig_values
    }
```
Working on a fix.

Reply:
Fix brought up some details with `as_list_col = TRUE`.  I'm not sure that `as_list_col = TRUE` even serves a purpose in `epi_slide` since I think it essentially requires every list entry (after any recycling) to be `vctrs::vec_size` 1.  Relates to #261 in another way.